Documentation/btrfs-check.rst

```diff
@@ -19,6 +19,12 @@ by the option *--readonly*.
 
 :command:`btrfsck` is an alias of :command:`btrfs check` command and is now deprecated.
 
+.. note::
+   Even though the filesystem checker requires a device argument, it scans for all
+   devices belonging to the same filesystem and may report metadata errors from other
+   devices that are correctable by :command:`btrfs scrub`. In this case, run scrub
+   first to ensure any correctable metadata errors are fixed to avoid false-positives.
```

Comment:
I do not think this note is needed.

Btrfs check handles metadata error just like kernel, if a mirror is corrupted but still have another good one available, btrfs check won't report it as an error. E.g:

```
Opening filesystem to check...
checksum verify failed on 30490624 wanted 0xcdcdcdcd found 0x86f5c8d8 << Mirror 1 corrupted for csum root
Checking filesystem on test.img
UUID: f01a583a-df1e-414b-a24c-7fe8bf2ef019
[1/8] checking log skipped (none written)
[2/8] checking root items
[3/8] checking extents
[4/8] checking free space tree
[5/8] checking fs roots
[6/8] checking only csums items (without verifying data)
[7/8] checking root refs
[8/8] checking quota groups skipped (not enabled on this FS)
found 147456 bytes used, no error found <<< Still no error.
total csum bytes: 0
total tree bytes: 147456
total fs tree bytes: 32768
total extent tree bytes: 16384
btree space waste bytes: 140595
file data blocks allocated: 0
 referenced 0
```

So there is no false positive. And `btrfs scrub` has a much higher bar to clear (RW mount), meanwhile btrfs check is always the most comprehensive check tool.

Reply:
This directly contradicts my experience when using btrfs-progs 6.6.3. In which version was this changed/fixed?

Reply:
IIRC it's always the case from the beginning.

Mind to give an example where btrfs-check is reporting recoverable corrupted mirrors as an critical error?
You can use `btrfs-map-logical` to find out the physical address of a metadata copy and corrupted it.

Reply:
The history of the problem is linked in the PR description. Briefly: a disk failed and temporarily dropped from the array. Therefore, the two disks were no longer in sync after rebooting. `btrfs check` claimed the filesystem was corrupted and printed messages to that end.

In consultation with @Zygo on IRC I manually verified by `btrfs inspect-internals` that `btrfs check` was reporting invalid node types/values that existed only on the bad disk. The kernel logs showed the kernel driver was noticing the invalid metadata nodes on the bad disk and switching to the good one, but `btrfs check` was claiming filesystem corruption even though the good disk was not corrupted and the filesystem metadata was fine. There was no indication by `btrfs check` that it was checking (only? due to PID load-balancing?) the bad device even when explicitly given the path to the good disk. Only after read-write `btrfs scrub` fixed up the metadata on the bad disk did `btrfs check` report no errors.

Reply:
Do you still have that fs?

I'm checking the `btrfs_read_extent_buffer()` function and still can not find out a good way where split-brain case can make btrfs-check to go the bad copy without trying the good one.

I think it's more like a bug other than an expected behavior.

Reply:
> Do you still have that fs?

I do not. However, until something is done to fix the lost writes problem (like [this MVP idea from Oct 2](https://lore.kernel.org/linux-btrfs/b5b3df31-627a-403c-af4d-7495e7c28bad@zetafleet.com/)), it seems like a synthetic reproducer should be trivial: create a raid1 fs from two disks, forcibly remove one disk, ensure btrfs continues writing to the remaining disk for a little bit, then unmount the fs, reconnect the forcibly removed disk, and remount.

> I'm checking the `btrfs_read_extent_buffer()` function and still can not find out a good way where split-brain case can make btrfs-check to go the bad copy without trying the good one.

If the problem can’t be found in the latest version then maybe it was fixed between 6.6.3 and current head? This was btrfs-progs 6.6.3 (which is [still the latest available packaged version for Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1074237)) on kernel 6.9.10.

> I think it's more like a bug other than an expected behavior.

Sure, this seems likely to me.